Support Union in TorchScript

ansley opened the pull request:
This PR is created to replace https://github.com/pytorch/pytorch/pull/53180 PR stack, which has all the review discussions. Reason for needing a replacement is due to a messy Sandcastle issue.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @cbalioglu @gcramer23 @ezyang @gchanan @agolynski @mrzzd

albanD commented on test/expect/TestScript.test_annot_ast_mypy_fn.expect:
```diff
@@ -6,4 +6,4 @@ foo(bool x, (Tensor, Tensor) y) -> ((bool, bool))
 foo(int[] x, (Tensor, Tensor) y) -> ((int[], int[]))
-foo(int? x, (Tensor, Tensor) y) -> ((int?, int?))
\ No newline at end of file
+foo(int? x, (Tensor, Tensor) y) -> ((int?, int?))
```
Why add new lines to these files?

albanD commented on test/test_public_bindings.py:
```diff
@@ -210,6 +210,7 @@ def test_no_new_bindings(self):
             "TupleType",
             "Type",
             "unify_type_list",
+            "UnionType",
```
Do we really want this to be in `torch.`?

albanD replied:
You need to add documentation to this if it is public API on `torch.` no?

suo replied:
cc @ansley can you comment?

ansley replied:
Thank you all for bringing this to my attention! I missed Alban's first comment, and the second one came after I had shipped the diff. I didn't want to revert it because it was high priority to get it out before the branch cut. I'm actively working on a cleanup diff now

albanD replied:
Thanks!

Commit:
Support Union in TorchScript (#64234)

Summary:
This PR is created to replace https://github.com/pytorch/pytorch/pull/53180 PR stack, which has all the review discussions. Reason for needing a replacement is due to a messy Sandcastle issue.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/64234

Reviewed By: gmagogsfm

Differential Revision: D30656444

Pulled By: ansley

fbshipit-source-id: 78d36669b7db30b5fefbd4f25d173c8c492e747d

eellison commented on test/test_jit.py:
```diff
@@ -8900,6 +8875,7 @@ def test_pack_unpack_state(self):
         torch.testing.assert_close(imported(x), x + torch.neg(torch.ones(3, 4, dtype=torch.float)))
 
     @unittest.skipIf(not TEST_MKL, "PyTorch is built without MKL support")
+    @unittest.skipIf(True, "Skipping while landing PR stack")
```
This is regressing coverage of core pytorch operators.. really not ideal to comment out other tests as part of landing a PR especially without notifying the people who added those tests.

eellison commented on test/test_ops.py:
```diff
@@ -684,6 +685,7 @@ class TestJit(JitCommonTestCase):
     #   and runtimes (eager, traced, scripted).
     # TODO WARNING: inplace x {traced, scripted} not currently tested
     @_variant_ops(op_db)
+    @unittest.skipIf(True, "Temporarily skipping while landing Union PR stack")
```
We rely on the OpInfos for a huge amount of testing of a number of essential systems. Turning them all off especially without looping anyone isn't great.

eellison reviewed (commented):
We're regressing coverage of at least operators in torch/functional.py and I dont know what's breaking in test/test_ops. I would normally say we should revert but it's been 6 days and that would just lead to more breakages. In the future we should be avoiding commenting out tests to land new PRs and at the very least loop relevant test owners in. cc @gmagogsfm 

I discovered this because all of the shape function tests I have added are now being completely skipped
